beancount-parser/src/lib.rs

```diff
@@ -1052,6 +1117,81 @@ mod tests {
         parse_ok!(posting, "Assets:Cash 200 XYZ {{}}");
     }
 
+    #[test]
+    fn pushtag() {
+        parse_ok!(pushtag, "pushtag #sometag\n");
+        parse_ok!(pushtag, "pushtag    #sometag\n");
+        parse_ok!(pushtag, "pushtag   #sometag  \n");
+        parse_fail!(pushtag, "pushtag\n");
+        parse_fail!(pushtag, "pushtag #goodtag #badtag\n");
+    }
+
+    #[test]
+    fn poptag() {
+        parse_ok!(poptag, "poptag #sometag\n");
+        parse_ok!(poptag, "poptag    #sometag\n");
+        parse_ok!(poptag, "poptag   #sometag  \n");
+        parse_fail!(poptag, "poptag\n");
+        parse_fail!(poptag, "poptag #goodtag #badtag\n");
+    }
+
+    #[test]
+    fn test_push() {
+        let mut state = ParseState::new();
+        state.push_tag("sometag");
+        assert_eq!(1, state.pushed_tags.len());
+        assert_eq!(Some(&1), state.pushed_tags.get("sometag"));
+        state.push_tag("othertag");
+        assert_eq!(2, state.pushed_tags.len());
+        assert_eq!(Some(&1), state.pushed_tags.get("othertag"));
+        assert_eq!(Some(&1), state.pushed_tags.get("sometag"));
+        state.push_tag("sometag");
+        assert_eq!(2, state.pushed_tags.len());
+        assert_eq!(Some(&2), state.pushed_tags.get("sometag"));
+    }
+
+    #[test]
+    fn test_pop() {
+        let mut state = ParseState::new();
+        assert!(state.pop_tag("sometag").is_err());
+        state.push_tag("sometag");
+        state.push_tag("sometag");
+        assert_eq!(1, state.pushed_tags.len());
+        assert_eq!(Some(&2), state.pushed_tags.get("sometag"));
+        assert!(state.pop_tag("sometag").is_ok());
+        assert_eq!(1, state.pushed_tags.len());
+        assert_eq!(Some(&1), state.pushed_tags.get("sometag"));
+        assert!(state.pop_tag("sometag").is_ok());
+        assert_eq!(0, state.pushed_tags.len());
+        assert_eq!(None, state.pushed_tags.get("sometag"));
+    }
+
+    fn get_sorted_tags<'a>(state: &'a ParseState) -> Vec<&'a str> {
+        let mut tags = state.get_tags().map(|a| *a).collect::<Vec<&'a str>>();
+        tags.sort();
+        tags
+    }
+
+    #[test]
+    fn test_iter() {
+        let mut state = ParseState::new();
+
+        assert!(get_sorted_tags(&state).is_empty());
+        state.push_tag("sometag");
+        assert_eq!(vec!["sometag"], get_sorted_tags(&state));
+        state.push_tag("sometag");
+        assert_eq!(vec!["sometag"], get_sorted_tags(&state));
+        state.push_tag("othertag");
+        assert_eq!(vec!["othertag", "sometag"], get_sorted_tags(&state));
+        assert!(state.pop_tag("sometag").is_ok());
+        assert_eq!(vec!["othertag", "sometag"], get_sorted_tags(&state));
+        assert!(state.pop_tag("sometag").is_ok());
+        assert_eq!(vec!["othertag"], get_sorted_tags(&state));
+        assert!(state.pop_tag("othertag").is_ok());
+        assert!(get_sorted_tags(&state).is_empty());
+        assert!(state.pop_tag("othertag").is_err());
+    }
+
```

Comment:
Adding a test that ensures transactions following a `pushtag` actually have that tag would be nice! Here's a test that you could use an example: https://github.com/twilco/beancount/blob/a1d948b09fdbd50a481e3839be4bc8b90c827c0a/beancount-parser/src/lib.rs#L1120#L1163